cpp/src/Slice/Parser.cpp

```diff
@@ -4388,13 +4376,31 @@ Slice::Operation::returnsClasses(bool includeTagged) const
 bool
 Slice::Operation::returnsData() const
 {
-    return returnType() || !_outParameters.empty() || !_throws.empty();
+    return !_returnValues.empty() || !_throws.empty();
 }
 
 bool
 Slice::Operation::returnsMultipleValues() const
 {
-    return _outParameters.size() + (returnType() ? 1 : 0) > 1;
+    return _returnValues.size() > 1;
+}
+
+bool
+Slice::Operation::hasOutParameters() const
+{
+    return _usesOutParameters;
+}
+
+bool
+Slice::Operation::hasTupleReturnType() const
+{
+    return !_usesOutParameters && _returnValues.size() > 1;
+}
+
+bool
+Slice::Operation::hasSingleReturnType() const
```

Comment:
Doesn't `_returnValues.size() == 1` implies a single return type, do we really need this methods? seems we can just check `returnValues()` size

Reply:
Not at the parsing level, since we combine return values and out parameters into a single list, so:
```
int op(out int j);
```
would have `returnValues.size() == 2` even though there's only a single return type. This might not matter for the language mappings where we'll combine out parameters and return tuples, but at the parsing level there are instances where we treat them differently.

For instance, with metadata, if there's a single return type, we apply the metadata to the return type and the operation, but if there's a tuple we dont:
```
[metadata] int op1(out int j);
[metadata] (int i1, int i2) op2();
```
Both of these cases would have `_returnValues.size() == 2`, but in the first case we need to apply the metadata to the `int` return type, in the second case we don't.